### ISSUE TITLE
 'void org.altbeacon.beacon.service.scanner.CycledLeScanner.stop()' on a null object reference

### ISSUE THREAD

**Issue by kababu007** [AltBeacon/android-beacon-library#758]
### Expected behavior
Should not crash when the library stops scanning for beacons.

### Actual behavior
Crash recorded:
java.lang.NullPointerException: Attempt to invoke virtual method 'void org.altbeacon.beacon.service.scanner.CycledLeScanner.stop()' on a null object reference
	at org.altbeacon.beacon.service.ScanJob.stopScanning()(unknown:169)
	at org.altbeacon.beacon.service.ScanJob.onStopJob()(unknown:161)
	at android.app.job.JobService$1.onStopJob()(JobService.java:76)
	at android.app.job.JobServiceEngine$JobHandler.handleMessage()(JobServiceEngine.java:117)
	at android.os.Handler.dispatchMessage()(Handler.java:106)
	at android.os.Looper.loop()(Looper.java:164)
	at android.app.ActivityThread.main()(ActivityThread.java:6651)
	at java.lang.reflect.Method.invoke()(Method.java:-2)
	at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run()(RuntimeInit.java:438)
	at com.android.internal.os.ZygoteInit.main()(ZygoteInit.java:810)

### Steps to reproduce this behavior
Not able to replicate

### Mobile device model and OS version
Device Model  LG-Q710AL
Device Name   LGE
Device Type    LG Device
OS Version     8.1.0

### Android Beacon Library version
2.15.2

IMPORTANT:  This forum is reserved for feature requests or reproducible bugs with the library itself.  If you need help with using the library with your project, please open a new question on StackOverflow.com.

**Comment by kababu007** [comment]
Just verified the crash logs and it happens only with version 2.15.2

**Comment by ryot** [comment]
One of our partners is also reporting this crash after updating to 2.15.2, sourced from a OnePlus One running Android 8.1
```
java.lang.RuntimeException:
at android.app.job.JobServiceEngine$JobHandler.handleMessage (JobServiceEngine.java:121)
at android.os.Handler.dispatchMessage (Handler.java:106)
at android.os.Looper.loop (Looper.java:164)
at android.app.ActivityThread.main (ActivityThread.java:6494)
at java.lang.reflect.Method.invoke (Method.java)
at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run (RuntimeInit.java:440)
at com.android.internal.os.ZygoteInit.main (ZygoteInit.java:807)
Caused by: java.lang.NullPointerException:
at org.altbeacon.beacon.service.ScanJob.stopScanning (ScanJob.java:169)
at org.altbeacon.beacon.service.ScanJob.onStopJob (ScanJob.java:161)
at android.app.job.JobService$1.onStopJob (JobService.java:76)
at android.app.job.JobServiceEngine$JobHandler.handleMessage (JobServiceEngine.java:117)
at android.os.Handler.dispatchMessage (Handler.java:106)
at android.os.Looper.loop (Looper.java:164)
at android.app.ActivityThread.main (ActivityThread.java:6494)
at java.lang.reflect.Method.invoke (Method.java)
at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run (RuntimeInit.java:440)
at com.android.internal.os.ZygoteInit.main (ZygoteInit.java:807)
```

**Comment by davidgyoung** [comment]
Yep, the problem is that this code was added in 2.15.2 to protect against a OutOfMemoryException:

        if (mScanHelper.getCycledScanner() == null) {
            try {
                mScanHelper.createCycledLeScanner(mScanState.getBackgroundMode(), null);
            }
            catch (OutOfMemoryError e) {
                LogManager.w(TAG, "Failed to create CycledLeScanner thread.");
                return false;
            }
        }

But that code can leave `mScanHelper.getCycledLeScanner()` as null causing a NullPointerException when the onStopJob executes this code:

    private void stopScanning() {
        mInitialized = false;
        mScanHelper.getCycledScanner().stop();
        mScanHelper.getCycledScanner().destroy();
        LogManager.d(TAG, "Scanning stopped");
    }

The solution is to protect against this with:

    private void stopScanning() {
        mInitialized = false;
        CycledScanner scanner = mScanHelper.getCycledScanner();
        if (scanner != null) {
            scanner.stop();
            scanner.destroy();
        }
        LogManager.d(TAG, "Scanning stopped");
    }

I'm not sure why nobody who tested this solution saw this same NullPointerException -- but it seems it should have always happened.

**Comment by kababu007** [comment]
Hi Team - Do you have plans to release the fix soon?

**Comment by davidgyoung** [comment]
The last bug fix release 2.15.2 was two weeks ago.  Unless there is a critical bug, we won't do another one for another two weeks, so we can focus on getting other fixes in.  If you can make a case that this is critical (e.g. affecting app functionality vs. just showing up in crash reports) then it may be possible to expedite.

**Comment by kababu007** [comment]
ok Understood. I think we see very less occurrence of this crash for a week now. But we will keep watching this space for the fixed release.

**Comment by davidgyoung** [comment]
Fixed in #766
